app/services/service_limits.rb

```diff
@@ -0,0 +1,124 @@
+# ServiceLimits
+#
+# This class is the guard check for the HighlightsController CRUD actions
+class ServiceLimits
+  MAX_HIGHLIGHTS_PER_SOURCE_PER_USER = 300
+  MAX_HIGHLIGHTS_PER_USER            = 20_000
+  MAX_ANNOTATION_CHARS_PER_USER      = 300_000
+  MAX_CHARS_PER_ANNOTATION           = 1_000
+
+  class ServiceLimitsError < StandardError; end
+
+  class ExceededMaxHighlightsPerUser < ServiceLimitsError
+    def message
+      "Exceeded (#{MAX_HIGHLIGHTS_PER_USER} max highlights per user"
+    end
+  end
+
+  class ExceededMaxHighlightsPerUserPerSource < ServiceLimitsError
+    def message
+      "Exceeded (#{MAX_HIGHLIGHTS_PER_SOURCE_PER_USER} max highlights per user per source"
+    end
+  end
+
+  class ExceededMaxAnnotationCharsPerUser < ServiceLimitsError
+    def message
+      "Exceeded (#{MAX_ANNOTATION_CHARS_PER_USER} max annotation chars per user"
+    end
+  end
+
+  class ExceededMaxAnnotationChars < ServiceLimitsError
+    def message
+      "Exceeded (#{MAX_CHARS_PER_ANNOTATION} max chars per annotation"
+    end
+  end
+
+  def initialize(user_id:)
+    @user_id = user_id
+  end
+
+  def with_create_protection
+    ActiveRecord::Base.transaction do
+      yield(build_user).tap do |model|
+        raise ActiveRecord::RecordInvalid, 'Block did not yield an active record model' unless model.is_a?(Highlight)
+        raise ActiveRecord::RecordInvalid, 'Model errors detected' if model.errors.present?
+
+        track_and_validate_max_highlights(model)
+        track_and_validate_annotation_chars(model) if model.annotation.present?
+      end
+    end
+  end
+
+  def with_update_protection(prev_annotation_length:)
+    ActiveRecord::Base.transaction do
+      yield.tap do |model|
+        raise ActiveRecord::RecordInvalid, 'Block did not yield an active record model' unless model.is_a?(Highlight)
+        raise ActiveRecord::RecordInvalid, 'Model errors detected' if model.errors.present?
+
+        reset_max_for_annotations(user: build_user, by: -prev_annotation_length) if prev_annotation_length.present?
+        track_and_validate_annotation_chars(model) if model.annotation.present?
+      end
+    end
+  end
+
+  def with_delete_reincrement
+    ActiveRecord::Base.transaction do
+      yield.tap do |model|
```

Comment:
Need check that model is a highlight?  like have above.  Need check that deletion actually happened? (no errors, etc)